### PR TITLE
boards: shields: add shield directory to DTS_ROOT

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Treat shield directories containing a dts subdirectory as a DTS root directory.

This allows shield specific devicetree include files and bindings to be stored under the shield dts subdirectory.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

### CODE REVIEW

File: cmake/app/boilerplate.cmake [hunk 1]
```diff
@@ -335,6 +335,14 @@ foreach(root ${BOARD_ROOT})
         ${shield_dir}/${s_dir}/dts_fixup.h
         )
 
+      # search for shield dts root directory
+      if(EXISTS ${shield_dir}/${s_dir}/dts AND IS_DIRECTORY ${shield_dir}/${s_dir}/dts)
```
Comment (tejlmand) [inline cmake/app/boilerplate.cmake:339]: The test for existence of the dts folder in a DTS_ROOT is also done here:
https://github.com/zephyrproject-rtos/zephyr/blob/09e9736b4b1a3e8c4bb572a5fe544956800990b9/cmake/dts.cmake#L90-L97
https://github.com/zephyrproject-rtos/zephyr/blob/09e9736b4b1a3e8c4bb572a5fe544956800990b9/cmake/dts.cmake#L107-L109

so we don't need to test the existence of the `dts` folder here.

Removing `EXISTS` and `IS_DIRECTORY` here ensures that adding shield dir to the list of dts roots behaves the same as any other dts root (board, application, etc.).
It also has the benefit that we only need to maintain those checks at a single location.

---

To simplify, and try to keep names aligned, then maybe we should create a list containing shield dirs, like:
```
list(APPEND SHIELD_DIRS ${shield_dir}/${s_dir})
```
The order in that list should of course match the order in the `SHIELD_LIST` so they can be used together.

Having such a list would be similar the the `BOARD_DIR` variable, and using the name `SHIELD_DIRS` indicates a variable accessible in whole build system (I know we have not been good at keeping variable names consistent).

The `SHIELD_DIRS` can then be added to `DTS_ROOT` as other default roots, but will also be available for other uses if needed.

File: cmake/dts.cmake [hunk 2]
```diff
@@ -34,6 +34,7 @@ list(APPEND
   DTS_ROOT
   ${APPLICATION_SOURCE_DIR}
   ${BOARD_DIR}
+  ${shield_dts_roots}
```
Comment (tejlmand) [inline cmake/dts.cmake:37]: If you accept the other suggestion, this would become:

```suggestion
  ${SHIELD_DIRS}
```

**Review by tejlmand — CHANGES_REQUESTED** [review]
Thanks for this.
Looks very useful, but I think this needs to go into documentation somewhere.

File: cmake/dts.cmake (hunk 2 above)
Comment (henrikbrixandersen) [reply cmake/dts.cmake:37]: @tejlmand That's a good idea, thanks. I am blocked, however, by #30553.

File: cmake/app/boilerplate.cmake (hunk 1 above)
Comment (henrikbrixandersen) [reply cmake/app/boilerplate.cmake:339]: @tejlmand I see the idea behind having a `SHIELD_LIST` variable, but shouldn't its order follow `SHIELD` and not `SHIELD_LIST`? This is how `shield_dts_files`, `shield_dts_fixups`, and `shield_conf_files` are sorted today, and provides the user the possibility of controlling the evaluation order.

Comment (tejlmand) [reply cmake/app/boilerplate.cmake]: > This is how shield_dts_files, shield_dts_fixups, and shield_conf_files are sorted today, and provides the user the possibility of controlling the evaluation order.

good point, let's not do any changes in this PR that could impact this order.

**Review by mbolivar-nordic — APPROVED** [review]

File: cmake/app/boilerplate.cmake [hunk 3]
```diff
@@ -346,6 +346,11 @@ foreach(root ${BOARD_ROOT})
         ${SHIELD_DIR_${s}}/dts_fixup.h
         )
 
+      list(APPEND
+        SHIELD_DIRS
+        ${shield_dir}/${s_dir}
```
Comment (tejlmand) [inline cmake/app/boilerplate.cmake:351]: This must be updated to:
```suggestion
        ${SHIELD_DIR_${s}}
```

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: cmake/app/boilerplate.cmake (hunk 3 above)
Comment (henrikbrixandersen) [reply cmake/app/boilerplate.cmake]: @tejlmand Fixed, thanks.

**Review by tejlmand — APPROVED** [review]

**Pull request merged**